src/pykeen/sampling/basic_negative_sampler.py

```diff
@@ -39,23 +39,48 @@ def sample(self, positive_batch: torch.LongTensor) -> torch.LongTensor:
         # Bind number of negatives to sample
         num_negs = positive_batch.shape[0]
 
-        # Equally corrupt head and tail
-        split_idx = num_negs // 2
+        # Equally corrupt all sides
+        split_idx = num_negs // len(self._corruption_indices)
 
         # Copy positive batch for corruption.
         # Do not detach, as no gradients should flow into the indices.
         negative_batch = positive_batch.clone()
 
         # Sample random entities as replacement
-        negative_entities = torch.randint(high=self.num_entities - 1, size=(num_negs,), device=positive_batch.device)
-
-        # Replace heads – To make sure we don't replace the head by the original value
-        # we shift all values greater or equal than the original value by one up
-        # for that reason we choose the random value from [0, num_entities -1]
-        filter_same_head = (negative_entities[:split_idx] >= positive_batch[:split_idx, 0])
-        negative_batch[:split_idx, 0] = negative_entities[:split_idx] + filter_same_head.long()
-        # Corrupt tails
-        filter_same_tail = (negative_entities[split_idx:] >= positive_batch[split_idx:, 2])
-        negative_batch[split_idx:, 2] = negative_entities[split_idx:] + filter_same_tail.long()
+        if 0 in self._corruption_indices or 2 in self._corruption_indices:
+            negative_entities = torch.randint(
+                high=self.num_entities - 1,
+                size=(num_negs,),
+                device=positive_batch.device,
+            )
+
+        # Sample random relations as replacement, if requested
+        if 1 in self._corruption_indices:
+            negative_relations = torch.randint(
+                high=self.num_relations - 1,
```

Comment:
is `high` inclusive?

Reply:
No it is exclusive. But there is an additional shift further in the code, since we want to sample uniformly from `{0, ..., max_id-1} \ {specific_id}`.

Reply:
No

Reply:
Thanks for the explanation @mberr . would be good for that to be in the code as well.

Reply:
Resolved in bf50831d